architecture/crds.md

```diff
@@ -0,0 +1,324 @@
+
+- A chart author: A person filling this role _creates_ and _maintains_ Helm charts
+- A Helm user: This person installs and manages instances of charts
+
+### The Chart Author Role
+
+We assume that a _chart author_ has three specific areas of domain knowledge:
+
+1. How to model applications in Kubernetes
+2. How to create Helm charts, including author templates
+3. How the application they are packaging works
+
+In this case, we assume the chart author role includes knowledge of Kubernetes kinds, Helm template commands, and so on.
+
+We do not assume that chart authors have knowledge about the clusters into which their charts are deployed, knowledge of the source code for the packages they install, or knowledge of the extended toolchains that their users have employed. Furthermore, we do not assume that chart authors will always follow best practices or accommodate use cases that may be important to some class of users. In fact, the Helm security model urges us to include bad actors in the class of chart authors. (That is, there is or may be a small subclass of chart authors that have intentions counter to the desires of their target Helm users.)
+
+### The Helm User Role
+
+Our assumptions about _Helm users_ are more modest. We do not assume they know much about Kubernetes or Helm--perhaps only enough to follow the Quickstart guide for Helm. With the base Helm user, we do not assume that they know what a Pod is, let alone a CRD. While we do assume that they know a little about YAML, we make no assumptions that they know about the Kubernetes flavor of YAML.
+
+### The Importance of This Distinction
+
+Over time, our assumptions have shown themselves true. Many (perhaps even most) Helm users are new to Kubernetes, and we hear repeatedly that people have learned Kubernetes via Helm. Our issue queue is replete with examples of people who have installed Helm charts in production, but who are not Kubernetes experts by any measure. Chart authoring, on the other hand, has remained the domain of experienced Kubernetes users, and the questions we receive from chart authors indicate a high degree of comfort with Kubernetes itself.
+
+Some have attempted to argue that "really," the chart developer and the Helm user are the same person -- that most of the time, people build their own charts. Our usage statistics show otherwise. The usage pattern we see most often with Helm is that a chart author is a different person than a Helm user. That is, most of the time, one group of persons creates charts, and another group of (non-overlapping) persons use the charts.
+
+As a consequence of this, we can assume that the _person using the chart does not know or understand the internals of the packages they install_. This is not merely a statement that they have not read the templates, but that they would not understand them even if they did, because they have not had to (nor should they have to) become fluent in the chart system to use the charts.
```

Comment:
I feel there's a kind-of gap here, in the error-handling case. If you install a chart, and it happens to overlap a resource unrelated, you're going to get an error that requires you to _at least_ know that the chart is creating a resource, and a resource of that name exists already for other reasons. The assumption given here would leave a user unable to resolve this problem.

So perhaps we can assume that "the person using the chart does not _initially_ know or understand the internals of the packages they install". And they probably never have to, but they may need to be able to _list_ the templates (or the generated resources, rather), even if they contents are opaque magic, so they can reason about the conflict.

I raise this because dealing with CRDs rationally in anything except the singleton-operator case is going to require at least that level of sophistication: "There are CRDs in the cart, there are CRDs with the same names on the cluster, and Helm is advising me X: I need to make a choice", even if the choice is "seek help", there needs to be enough information/sophistication that help _can_ be sought effectively.

Reply:
@TBBle providing more information to troubleshoot makes sense. But, I would suggest that information needs more context and detail than raw manifests.

Consider the case of a PostgreSQL operator being installed. The person who created the chart needs to know how to operationalize PostgreSQL, the operation of operators, and Kubernetes. The person who installed the chart needs to know none of these things. To troubleshoot a problem installing the chart and giving them raw Kubernetes manifests would require a HUGE learning curve to understand them.

I think there is a feature idea here about helping people deal with CRD issues that isn't material to this document. One about giving people more information to help them troubleshoot.

Reply:
Yeah, that makes sense. I don't think the solution is to "dump the whole YAML blob on them". My point was more "you have to be twice as smart to debug it as you were to ~~write the code~~ install the chart" and that there's going to be a reasonable userbase who aren't expert in YAML manifests, or even really aware to look at them, but are going to build by necessity more expertise than "just running `helm install`".

The only reason I raise it here is that I think dealing with CRDs is likely to be the major source of that suddenly-necessary expertise, and so I don't want to see this bare-minimum expertise assumption lead to closing off options prematurely.

Happy to punt it as separate to this document, the relevant aspects of this are separately discussed elsewhere anyway.